### PR TITLE
List Widget Phase 2

### PR DESCRIPTION
**Pull request by jsartisan**
This PR makes input widgets work in list widget

Fixes #2389 

## Test coverage results :test_tube:
<details><summary>:green_circle: Total coverage has increased</summary>

    // Code coverage diff between base branch:release and head branch: feature/list-widget-phase-2 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :green_circle: | total | 53.3 **(0.01)** | 35.26 **(-0.01)** | 31.92 **(0.04)** | 53.81 **(0.01)**
 :red_circle: | app/client/src/components/designSystems/appsmith/ContainerComponent.tsx | 92.86 **(0)** | 55 **(-0.56)** | 100 **(0)** | 92.86 **(0)**
 :green_circle: | app/client/src/components/editorComponents/ResizableComponent.tsx | 59.34 **(1.87)** | 22.95 **(0)** | 63.64 **(3.64)** | 58.43 **(1.96)**
 :green_circle: | app/client/src/pages/Editor/Explorer/Pages/PageContextMenu.tsx | 72.22 **(0.79)** | 60 **(0)** | 12.5 **(0)** | 74.29 **(0.76)**
 :red_circle: | app/client/src/pages/Editor/Explorer/Widgets/WidgetContextMenu.tsx | 31.58 **(-3.71)** | 0 **(0)** | 0 **(0)** | 32.43 **(-3.93)**
 :green_circle: | app/client/src/reducers/entityReducers/metaReducer.ts | 28.57 **(7.74)** | 0 **(0)** | 0 **(0)** | 28.57 **(7.74)**
 :green_circle: | app/client/src/resizable/index.tsx | 56.76 **(11.93)** | 41.67 **(6.67)** | 31.58 **(0)** | 58.73 **(14.05)**
 :red_circle: | app/client/src/utils/WidgetPropsUtils.tsx | 65.63 **(-1.04)** | 59.41 **(-2.58)** | 56.41 **(-2.49)** | 65.03 **(-1.07)**
 :red_circle: | app/client/src/utils/helpers.tsx | 51.63 **(-0.07)** | 22.54 **(-0.65)** | 19.23 **(-0.77)** | 45 **(-0.22)**
 :green_circle: | app/client/src/widgets/BaseWidget.tsx | 80.67 **(0.16)** | 41.86 **(0)** | 80 **(0)** | 79.65 **(0.19)**
 :green_circle: | app/client/src/widgets/FilepickerWidget.tsx | 39.33 **(1.4)** | 0 **(0)** | 18.52 **(3.14)** | 38.64 **(1.43)**
 :red_circle: | app/client/src/widgets/MetaHOC.tsx | 82.09 **(-0.72)** | 57.89 **(-0.93)** | 92.31 **(0)** | 82.76 **(-0.88)**
 :red_circle: | app/client/src/widgets/ListWidget/ListComponent.tsx | 66.67 **(-13.33)** | 61.11 **(11.11)** | 33.33 **(-16.67)** | 66.67 **(-13.33)**
 :green_circle: | app/client/src/widgets/ListWidget/ListWidget.tsx | 38.84 **(0.84)** | 22.86 **(0.03)** | 27.91 **(1.44)** | 38.16 **(1.06)**
 :sparkles: :new: | **app/client/src/widgets/ListWidget/derived.js** | **62.22** | **41.67** | **70** | **61.36**
 :sparkles: :new: | **app/client/src/widgets/ListWidget/parseDerivedProperties.ts** | **38.46** | **0** | **0** | **38.46**</details>

### CODE REVIEW

**Comment by jsartisan** [comment]
@prapullac Can we test this?

File: app/client/src/reducers/entityReducers/metaReducer.ts [hunk 1]
```diff
@@ -11,14 +12,14 @@ export const metaReducer = createReducer(initialState, {
...
   ) => {
-    const next = { ...state };
-    let widgetMetaProps: Record<string, any> = next[action.payload.widgetId];
-    if (widgetMetaProps === undefined) {
-      widgetMetaProps = {};
-      next[action.payload.widgetId] = widgetMetaProps;
-    }
-    (widgetMetaProps as Record<string, any>)[action.payload.propertyName] =
-      action.payload.propertyValue;
+    const next = cloneDeep(state);
```
[the remark this answers is not in the document]
Comment (riodeuno) [reply app/client/src/reducers/entityReducers/metaReducer.ts:15]: @hetunandu will the above comment work?

File: app/client/src/widgets/BaseWidget.tsx [hunk 2]
```diff
@@ -385,7 +385,8 @@ export interface BaseStyle {
 export type WidgetState = Record<string, unknown>;
 
 export interface WidgetBuilder<T extends WidgetProps, S extends WidgetState> {
-  buildWidget(widgetProps: T): JSX.Element;
+  // step 4 - accept options as 2nd param and use proper types for options
+  buildWidget(widgetProps: T, options?: any): JSX.Element;
```
Comment (riodeuno) [inline app/client/src/widgets/BaseWidget.tsx:389]: Let's see if `options` argument is necessary.

File: app/client/src/workers/DataTreeEvaluator.ts [hunk 3]
```diff
@@ -429,6 +429,7 @@ export default class DataTreeEvaluator {
             const widgetEntity = entity;
             const defaultPropertyMap = this.widgetConfigMap[widgetEntity.type]
               .defaultProperties;
+
```
Comment (riodeuno) [inline app/client/src/workers/DataTreeEvaluator.ts:432]: Let's avoid these types of changes.

File: app/client/src/widgets/ListWidget/ListWidget.tsx [hunk 4]
```diff
@@ -49,21 +49,96 @@ class ListWidget extends BaseWidget<ListWidgetProps<WidgetProps>, WidgetState> {
 
   static getDerivedPropertiesMap() {
     return {
+      items: `{{(() => {
+        let item = {};
+
+        Object.keys(this.template).map(widgetName => {
+          item[widgetName] = {...this.template[widgetName] };
```
Comment (riodeuno) [inline app/client/src/widgets/ListWidget/ListWidget.tsx:56]: Let's move it to a different js file, just like table widget and add tests for these.

File: app/client/src/reducers/entityReducers/metaReducer.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (hetunandu) [reply app/client/src/reducers/entityReducers/metaReducer.ts:15]: Yea it should. Main idea is to not mutate the old state

**Review by marks0351 — DISMISSED** [review]

**Review by marks0351 — APPROVED** [review]

**Review by somangshu — APPROVED** [review]

File: app/client/src/widgets/MetaHOC.tsx [hunk 5]
```diff
@@ -124,18 +124,32 @@ const withMeta = (WrappedWidget: typeof BaseWidget) => {
     handleUpdateWidgetMetaProperty() {
       const { executeAction, updateWidgetMetaProperty } = this.context;
       const { widgetId, widgetName } = this.props;
+      const metaOptions = this.props.__metaOptions;
```
Comment (Rishabh-Rathod) [inline app/client/src/widgets/MetaHOC.tsx:127]: Can you help me understand where we use meta options in List Widget? @jsartisan

**Pull request merged**